### ISSUE TITLE
[Sapling] If no change is given, this leaks that the output value is the same value as a previous output note

### ISSUE THREAD

**Issue by daira** [zcash/zcash#3606]
(or, sum of previous output notes in the case where there is more than one shielded input). This is mostly significant for unshielding transactions where the value is public, but it is also technically an information leak in the fully shielded case.

**Comment by daira** [comment]
Note that the current transaction builder only creates a change output if the change value would be nonzero:

https://github.com/zcash/zcash/blob/625797a0371cac7044a9eb2872ee3820728c1d7d/src/transaction_builder.cpp#L123-L129

This is a potential information leak, which does not only affect unshielding transactions. It should always create the change output.

**Comment by str4d** [comment]
I think all we need to do for this is replace the above with `if (change >= 0) {` (it's signed, though we have just handled the negative case, but this would mean no indentation changes). The rest of the `TransactionBuilder` code should work fine (it would be treated as a dummy output that happens to have a valid target address).

**Comment by daira** [comment]
The difference between that and an actual dummy output, is that the recipient will see it as a valid payment (of zero value) when scanning the blockchain. This has a (very small) performance cost. Note that scanning with a full viewing key will process the transaction anyway.

**Comment by SamsungGalaxyPlayer** [comment]
I agree with @str4d. Fully-shielded transactions should have at least 2 transactions outputs for the sake of reducing the impact of this metadata.

**Comment by gojomo** [comment]
If the newly-transparent value exactly correlates with some earlier known (sum of) transparent value, it strikes me that adding a decoy output of (secretly) 0-value doesn't help much – especially if the creation of such decoys is enforced and thus assumed by analysts. It's that act of cycling full amounts t->z(->z)->t that has inherent risks, and should be discouraged by education & software.

**Comment by SamsungGalaxyPlayer** [comment]
@gojomo this isn't only important for cases involving t addresses. It also has implications on the spending patterns of fully-shielded transactions.

**Comment by daira** [comment]
Consider unshielding transactions such as https://zcha.in/transactions/ec5e3c4b177be99d5ffe2e54b60656afe6202ceaede55776e0c666835200c944 .

Notice that in Sprout, we can't tell whether this transaction was also making a shielded transfer. We can *guess* that it isn't and that the sum of the shielded inputs was exactly 10.02396116 ZEC, but it is also quite possible that the shielded input was 10.02396116+x ZEC and the shielded output was x ZEC. But in Sapling as implemented in 2.0.1, we'll be able to exclude that possibility if it is not the case. This is obviously a regression and should be fixed.

A significant advantage we have in Sapling is that outputs are very cheap; it only takes about 0.2 seconds to make an output proof (I think this was on @str4d's laptop). So, a policy of always having at least two shielded outputs for not-fully-transparent transactions would not be ruled out on performance grounds, and is worth seriously considering. That's strictly speaking not in scope for this ticket and should be discussed on #3615, though.

**Comment by daira** [comment]
This should be fixed for 2.0.2; the desired behaviour change is clear and there's easily enough time to implement it.

**Comment by gojomo** [comment]
I agree the (z->t) unshielding case is *potentially* the most concerning, and clearly there's now *slightly* more info about the value of the z-input(s) in the Spaling case vs the Sprout case. 

But I'm still struggling to see a meaningful impact. The absolute worst case would be a (t->z) that is observed to create exactly one 10.02396116 ZEC zc-output, and then later a (z->t) that consumes exactly one 10.02396116 ZEC zc-input. (Cases where an adversary has visibility into some z-transactions, by being the counterparty or compromising other counterparties, seem largely analogous to this t->z, z->t cycle.) 

Such tight t->z->t cycles are a bad habit that should be discouraged in education & software. But even in that case, essentially all of the the metadata-harm flows from the close-value-correlation, or timing info, or traffic-analysis. A decoy 0-value output hardly changes those correlations - especially if the creation of such 0-value decoys is automatic, and thus presumed likely when the revealed values (or other observable properties) match up. 

But that decoy output does, per @daira's ["arguing the other side" observations](https://github.com/zcash/zcash/issues/3615#issuecomment-432503768) on #3615, add 948 bytes to the blockchain, an extra note commitment, and trial-decryption costs for every user who ever scans the blockchain. Those appear to me to be real costs for a hand-wavey, negligible benefit. That is, even if this is technically a "regression" with regard to what Sprout hid, the old behavior may have been overkill, and the new behavior a resource-savings with no real privacy cost. 

Knowing that even slight amounts of metadata can have non-intuitive privacy loss, I'm a little queasy to be arguing this, but would really like understand, and be convinced, if there's a case where a 1-vs-2 output-count creates the tangible privacy loss.

A related digression: if education & software prove insufficient to guide users away from engaging in these sorts of t->z->t bad-habits (as analyzed in [Quesnelle's "On the linkability of Zcash transactions"](https://arxiv.org/abs/1712.01210) or [Kappos/Yousaf/Maller/Meiklejohn's "An Empirical Analysis of Anonymity in Zcash"](https://arxiv.org/abs/1805.03180) should there be public-interest "Friends-of-Priva-Z" organization that generates chaff transactions, intentionally churning value by re-using idiosyncratic amounts observed in other public t->z shielding or z->t unshielding transactions? Such a strategy *might* better confuse analysis based on correlated-ZEC-values than automatic 0-value decoy outputs, for the same amount of synthetic output-creation. And thanks to the ZK-shielding, never actually commingles funds with other users under some shared spending authority – only mimics the shape of other public transactions.

**Comment by daira** [comment]
The Transaction builder implementation in Rust added in https://github.com/zcash/librustzcash/pull/92 implements the policy we decided on in this ticket, of always creating at least two Sapling outputs if there is at least one Sapling spend.

(This ticket should remain open because the C++ code does not implement this policy.)
